app/src/main/java/org/fossasia/pslab/communication/ScienceLab.java

```diff
@@ -167,17 +193,157 @@ private void runInitSequence(Boolean loadCalibrationData) throws IOException {
+
+
+            /* POLYNOMIAL DATA FOR CALIBRATION */
+            //byte[] polynomials = readBulkFlash(this.ADC_POLYNOMIALS_LOCATION, 2048);
+
+            ArrayList<Byte> polynomialsByteData = new ArrayList<>();
+            for (int i = 0; i < 2048 / 16; i++) {
+                byte[] temp = readFlash(ADC_POLYNOMIALS_LOCATION, i * 16);
+                for (byte a : temp) {
+                    polynomialsByteData.add(a);
+                }
+            }
+            String polynomialByteString = "";
+            for (int i = 0; i < 2048; i++) {
+                try {
+                    polynomialByteString += new String(new byte[]{polynomialsByteData.get(i)}, "UTF-8");
+                } catch (UnsupportedEncodingException e) {
+                    e.printStackTrace();
+                }
+            }
             // todo : change to "PSLab" after PSLab firmware is ready
-            if ("SEELablet".equals(new String(Arrays.copyOfRange(polynomials, 0, 9), "UTF-8"))) {
+            if ("SEELablet".equals(polynomialByteString.substring(0, 9))) {
                 Log.v(TAG, "ADC calibration found...");
                 this.calibrated = true;
-                byte[] adcShifts1 = readBulkFlash(this.ADC_SHIFTS_LOCATION1, 2048);
-                byte[] adcShifts2 = readBulkFlash(this.ADC_SHIFTS_LOCATION2, 2048);
+                ArrayList<Byte> adcShifts = new ArrayList<>();
+                for (int i = 0; i < 2048 / 16; i++) {
+                    byte[] temp = readFlash(ADC_SHIFTS_LOCATION1, i * 16);
+                    for (byte a : temp) {
+                        adcShifts.add(a);
+                    }
+                }
+                for (int i = 0; i < 2048 / 16; i++) {
+                    byte[] temp = readFlash(ADC_SHIFTS_LOCATION2, i * 16);
+                    for (byte a : temp) {
+                        adcShifts.add(a);
+                    }
+                }
+                String[] polySections = polynomialByteString.split("STOP");
+                String adcSlopeOffsets = polySections[0];
+                String dacSlopeIntercept = polySections[1];
+                String inlSlopeIntercept = polySections[2];
+
+                Map<String, ArrayList<Double[]>> polyDict = new LinkedHashMap<>();
+                String[] adcSlopeOffsetsSplit = adcSlopeOffsets.split(">|");
+                for (int i = 0; i < adcSlopeOffsetsSplit.length; i++) {
+                    if (i == 0) continue;
+                    String cals = adcSlopeOffsetsSplit[i].substring(5);
+                    polyDict.put(adcSlopeOffsetsSplit[i].substring(0, 3), new ArrayList<Double[]>());
+                    for (int j = 0; j < cals.length() / 16; j++) {
+                        Double[] poly = new Double[4];
+                        for (int k = 0; k < 4; k++) {
+                            poly[k] = ByteBuffer.wrap(cals.substring(k, k + 4).getBytes()).getDouble();
```

Comment:
Please check this line. (k, k + 4) I think it should be (k +4*j, k + 4*(j+1)). Please correct me if I am wrong.